### DISCUSSION TITLE
Has there been a change in the behaviour of classes inside modules recently?

### DISCUSSION THREAD

**Discussion by plastikfan** [PowerShell/PowerShell#17002] [Q&A]
Using classes in Powershell, is generally fraught with many issues, which requires quite a lot of hoop jumping to navigate successfully. One of the biggest problems is not being able to modify the definition of a class in the same powershell session (this is an over simplification of the problem, but will do for the purposes of this discussion). The way I mitigate this issue, is to restart the current powershell session before rebuilding the module with the new version of modified classes. In the past, the error that would be generated would be something like (`PSInvalidCastException`):

```
[-] Krayon.given: pair.and: pair created via default constructor.should: write pair 31ms (30ms|1ms)
 PSInvalidCastException: Cannot convert the "Krayon" value of type "Krayon" to type "Krayon".
 ArgumentTransformationMetadataException: Cannot convert the "Krayon" value of type "Krayon" to type "Krayon".
 at <ScriptBlock>, ..\github\PoSh\Krayola\Elizium.Krayola\Tests\Krayon.tests.ps1:21
```

However, I have noticed recently, that this behaviour seems to have changed. I now no longer see this error, which I initially thought was due to this issue being fixed. On further development and strange other behaviour, I have come to the conclusion that the class is not reloaded, ie the old implemenation is still in force. This causes a lot of confusion as to why certain functionality is not executing properly, because the old implementation of the class is still running instead of seeing this `PSInvalidCastException` error, so the error is being hidden, which arguably is worse than the prior behaviour. You still need to restart the session to get the new behaviour you're expecting to be present.

Has anyone else discovered this, or is it just me?

**Comment by SeeminglyScience** [comment]
It's always been sort of a toss up which behavior you get depending on what the changes you made were. Certain changes fail to create a new type at all, others create a new type but only in some scopes (which gives you the above error).

**Reply by plastikfan** [reply]
Its just that, I've never been able to modify a class's implentation without getting that cast error, until relatively recently. I can now modify a class without getting that error (even tho the result is not what I want, ie I want the new implementation butI get the old one). To me it feels like something behind the scenes has changed perhaps in release 7.2.1.

**Reply by SeeminglyScience** [reply]
I hear ya, if you can create minimal repro steps demonstrating a specific behavior change file an issue for sure. I can tell you for sure I've hit both behaviors with similar frequencies depending on what changes I made. For instance, iirc signature changes of an existing method is one change more likely to be ignored.
